scripts/ghidra/decompile-entrypoint.sh

```diff
@@ -43,152 +43,287 @@ EOF
-  if [[ -z "$INPUT_FILE" ]]; then
-    die "--input is required."
-  fi
-
-  if [[ -z "$COMMAND" ]]; then
-    die "--command is required."
-  fi
-
-  if [[ "$COMMAND" == "decompile" && -z "$FUNCTION_NAME" ]]; then
-    die "--function is required when --command decompile is specified."
-  fi
-
-  if [[ -z "$OUTPUT_FILE" ]]; then
-    die "--output is required."
-  fi
+	# Validate required arguments
+	if [[ -z "$INPUT_FILE" ]]; then
+		die "--input is required."
+	fi
+
+	if [[ -z "$COMMAND" ]]; then
+		die "--command is required."
+	fi
+
+	if [[ "$COMMAND" == "decompile" && -z "$FUNCTION_NAME" ]]; then
+		die "--function is required when --command decompile is specified."
+	fi
+
+	if [[ -z "$OUTPUT_FILE" ]]; then
+		die "--output is required."
+	fi
+}
+
+ARCHITECTURE=""
+# Create the Ghidra processor string (ARCHITECTURE) from attributes visible to 
+# readelf in the input binary that we intend to pass to Ghidra eventually.
+# The processor string tells Ghidra how to decompile the input binary, based on
+# the pspec and cspec files available to it. The string has four colon-sep 
+# parts (processor_name:endianness:mode:variant), the first
+# three of which relate to and help define which pspec Ghidra will use:
+#
+# The first part is the processor name (e.g., x86, ARM, MIPS).
+# The second part is the endianness (e.g., LE, BE).
+# The third part is the address size or mode (e.g., 32, 64, 16).
+# The fourth part is the compiler specification or a processor variant (e.g., 
+# default, windows, clang, or a custom variant for special processor features).
+# This defines which cspec Ghidra will use to decompile. This doesn't quite mean
+# "what compiler was used to compile the input binary", i.e., if a binary was
+# originally compiled with clang for linux, you'd still put default for
+# the variant string.
+function detect_processor {
+	local file_output=$(file "$INPUT_FILE")
+
+	local endianness="LE"
+	if [[ "$file_output" == *"big-endian"* ]]; then
+		endianness="BE"
+	fi
+
+	local variant="default"
+	if [[ "$file_output" == *"PE32"* ]]; then
+		variant="windows"
+	elif [[ "$file_output" == *"Mach-O"* ]]; then
+		variant="clang"
+	fi
+
+	local mode="32"
+	# this grep will take only the first match
+	local processor_name=$(echo "$file_output" | grep -o -E \
+		'x86-64|Intel 80386|ARMv[0-9]+|armv[0-9]+|ARM aarch64|ARM|AArch64|MIPS|PowerPC|AVR|MSP430|8051|68k|SPARC|RISC-V|Xtensa|CR16C|Z80|6502|PIC')
+	case "$processor_name" in
+		"x86-64")
+			processor_name="x86"
+			mode="64"
+			;;
+		"Intel 80386")
+			processor_name="x86"
+			;;
+		"ARM aarch64"|"AArch64")
+			processor_name="AARCH64"
+			mode="64"
+			variant="v8A"
+			;;
+		ARM*|armv*)
+			variant="v7"
+			processor_name="ARM"
+
+			local arm_attrs=$(readelf -A "$INPUT_FILE" 2>/dev/null)
+			if [[ "$arm_attrs" == *"Tag_CPU_arch: v8"* ]]; then
+				variant="v8"
+			elif [[ "$arm_attrs" == *"Cortex-M"* ]] || readelf -S "$INPUT_FILE" | grep -q -E "\.text.*08000000|\.data.*20000000"; then
```

Comment:
```
grep -q -E "\.text.*08000000|\.data.*20000000"
```
Curious if it is this standard check Cotex (text and data region addresses)?

Reply:
I believe Cortex-M SRAM should always start at 0x20000000 by spec and have left an explanatory comment to that effect... however, 0x80000000 is only the start of Flash for some Cortex-M, so I removed that 😅 

I think if attributes are present `Microcontroller` should generally show up - it's just, if attributes are stripped for whatever reason, it won't show up (nor will `Cortex-M`, since attributes would be stripped), and Cortex-M is a little bit weird since we're in Thumb instructions then.